packages/backend/src/matter/behaviors/thermostat-server.ts

```diff
@@ -1,18 +1,246 @@
-) {
-  if (supportsCooling && supportsHeating) {
-    return HeatingAndCoolingThermostatServer;
-  } else if (supportsHeating) {
-    return HeatingThermostatServer;
-  } else if (supportsCooling) {
-    return CoolingThermostatServer;
-  } else {
-    return undefined;
+import { ThermostatServer as Base } from "@matter/main/behaviors";
+import { Thermostat } from "@matter/main/clusters";
+import { HomeAssistantBehavior } from "../custom-behaviors/home-assistant-behavior.js";
+import {
+  ClimateDeviceAttributes,
+  ClimateHvacAction,
+  ClimateHvacMode,
+  HomeAssistantEntityState,
+} from "@home-assistant-matter-hub/common";
+import { Behavior } from "@matter/main";
+import _ from "lodash";
+import { ClusterType } from "@matter/main/types";
+
+const FeaturedBase = Base.with("Heating", "Cooling", "AutoMode");
+
+export class ThermostatServerBase extends FeaturedBase {
+  declare state: ThermostatServerBase.State;
+  declare internal: ThermostatServerBase.Internal;
+
+  override async initialize() {
+    await super.initialize();
+    const homeAssistant = await this.agent.load(HomeAssistantBehavior);
+    Object.assign(
+      this.internal,
+      this.getState(
+        homeAssistant.state
+          .entity as HomeAssistantEntityState<ClimateDeviceAttributes>,
+      ),
+    );
+
+    const updatedState: Partial<ThermostatServerBase.State> = {
+      localTemperature: this.internal.currentTemperature,
+      systemMode: this.internal.systemMode,
+      // thermostatRunningMode: this.internal.runningMode,
+      ...(this.features.heating
+        ? {
+            occupiedHeatingSetpoint: this.internal.targetTemperature,
+            minHeatSetpointLimit: this.internal.minTemperature,
+            maxHeatSetpointLimit: this.internal.maxTemperature,
+          }
+        : {}),
+      ...(this.features.cooling
+        ? {
+            occupiedCoolingSetpoint: this.internal.targetTemperature,
+            minCoolSetpointLimit: this.internal.minTemperature,
+            maxCoolSetpointLimit: this.internal.maxTemperature,
+          }
+        : {}),
+
+      controlSequenceOfOperation:
+        this.features.cooling && this.features.heating
+          ? Thermostat.ControlSequenceOfOperation.CoolingAndHeating
+          : this.features.cooling
+            ? Thermostat.ControlSequenceOfOperation.CoolingOnly
+            : Thermostat.ControlSequenceOfOperation.HeatingOnly,
+    };
+
+    Object.assign(this.state, updatedState);
+
+    this.events.occupiedHeatingSetpoint$Changed.on(
+      this.targetTemperatureChanged.bind(this),
+    );
+    this.events.systemMode$Changed.on(this.systemModeChanged.bind(this));
+    homeAssistant.onUpdate((s) => this.update(s));
+  }
+
+  private async update(state: HomeAssistantEntityState) {
+    Object.assign(
+      this.internal,
+      this.getState(state as HomeAssistantEntityState<ClimateDeviceAttributes>),
+    );
+    const actualState = this.endpoint.stateOf(ThermostatServerBase);
```

Comment:
This endpoint does not support thermostat. This is because `ThermostatServer` (line 244 of this file) is added to the device, but `ThermostatServer.supports(ThermostatBaseServer)` returns `false`.